### ISSUE TITLE
`no-return-assign` behavior changed with arrow functions

### ISSUE THREAD

**Issue by feross** [eslint/eslint#5150]
Using the latest eslint (2.0.0-rc.0).

Rule:

```json
{
  "no-return-assign": 2
}
```

Code:

```js
var data = ''
process.stdin.on('data', buffer => data += buffer.toString())
```

Produces the error:

```
  2:26  error  Arrow function should not return assignment  no-return-assign
```

The rule is functioning correctly, because there's an implicit return with arrow functions. But I'm wondering if there isn't a way to keep the rule enabled but accommodate this kind of usage with arrow functions.

The intent of the rule is, primarily, to catch cases where the user intended a comparison but accidentally did an assignment. There's no chance of that in this example because it's `+=` not `=`, yet this is still considered an error because there's an assignment happening.

Possible ways to accommodate this:

- add an `except-arrow-function` option, to disable checking on arrow functions
- add a way to specify only checking for assignments with `=`

**Comment by eslintbot** [comment]
@feross Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
If we omitted arrow functions, would it just be for concise bodies? Or also for block bodies?

**Comment by michaelficarra** [comment]
Since you're not using the return value here, it'd be appropriate for you to put the body in curlies. :-1: This rule is doing exactly what was intended.

**Comment by platinumazure** [comment]
@michaelficarra I agree with your reasoning but I can see why it could be confusing or frustrating for users since there is no way to "omit the return" without going to braces (and the rule messaging doesn't make that clear either). Surely this still merits a change of some kind? Either:

1. Change the messaging for concise arrow functions
1. Allow an option to suppress warnings on concise arrow functions

**Comment by feross** [comment]
An option to suppress warnings on concise arrow functions would be perfect.

**Comment by despairblue** [comment]
No idea if that is important for this, but using brackets also suppresses the warning, although it's still returning the assignment.
```js
// Arrow function should not return assignment
const a = (v) => v.b = 1
// passes
const a = (v) => (v.b = 1)
```

**Comment by michaelficarra** [comment]
@despairblue That's an intentional exception to this rule.

**Comment by nzakas** [comment]
I think that adding an exception just for arrow functions, or just for arrow functions with concise bodies, is a bit too specific an exception. We do have ways around this now, both by adding parentheses, adding braces, or using `/*eslint-disable*/`. I think @michaelficarra is correct in that this is really the type of nonobvious issue that the rule is trying to flag.

**Comment by feross** [comment]
@nzakas I get where you're coming from. That's what the rule is for.

This could be an accidental return:

```js
const a = (v) => v.b = 1
```

But this is almost certainly intentional:

```js
const a = (v) => v.b += 1
```

In any case, I don't feel too strongly about this. Feel free to close if this is out of the question.

**Comment by nzakas** [comment]
At the moment, it doesn't seem like there's much support for adding an exception to the rule, so closing (as always, we'll monitor to see if similar issues pop up in the future.)

**Comment by scottohara** [comment]
Please note that some of the suggested workarounds (adding parens/curlies around the expression) may conflict with other rules, such as `no-extra-parens`, `brace-style` (1tbs) and/or `semi`.

**Example:**
```
/*eslint no-return-assign: 2,
         no-extra-parens: 2,
         brace-style: 2,
         semi: 2
*/

somePromise.then(foo => foo.bar = 1);
```

As expected, the above code fails with:

```
error  Arrow function should not return assignment  no-return-assign
```

**Workaround 1:**  Wrap the assignment expression in parens:

```
/*eslint no-return-assign: 2,
         no-extra-parens: 2,
         brace-style: 2,
         semi: 2
*/

somePromise.then(foo => (foo.bar = 1));
```

We now fail with:

```
error  Gratuitous parentheses around expression     no-extra-parens
```

**Workaround 2:** Wrap the assignment expression in curlies:

```
/*eslint no-return-assign: 2,
         no-extra-parens: 2,
         brace-style: 2,
         semi: 2
*/

somePromise.then(foo => {foo.bar = 1});
```

We now fail with:

```
error  Statement inside of curly braces should be on next line  brace-style
error  Missing semicolon                                        semi
```

The `semi` error _could_ be mitigated by including `omitLastInOneLineBlock: true`; but it would still violate the 1tbs brace style

**Workaround 3:** Convert from concise body to full block body:

```
/*eslint no-return-assign: 2,
         no-extra-parens: 2,
         brace-style: 2,
         semi: 2
*/

somePromise.then(foo => {
  foo.bar = 1;
});
```

Success!   Although I would argue that the arrow function has lost some of it's original beauty here.

Based on this, I don't think any of the suggested workarounds (except perhaps temporarily disabling, via `eslint-disable`) are particularly appealing, considering their impact on other rules.

If you're still interested in votes for some kind of exception to the rule; consider this my +1.

**Comment by crookedneighbor** [comment]
I would also like an exception to this rule for arrow functions.

**Comment by aarongreenwald** [comment]
I'm going to add to the chorus requesting an exception (and if it would be accepted, I'd be willing to work on a PR to get it done). 

I like having `no-return-assign`, but I also want to error when braces are used in arrow functions that contain only one line. And I don't want to add visual weight to my code. For me, one of the key benefits of arrow functions is the ability to write clean, concise code:

```
someArray.forEach(item => item.prop = 'val')
```

Yes, I'm technically returning an assignment, but it's a common pattern that's not a mistake. If the spirit of `no-return-assign` is to prevent accidents, I think the rule should be limited to where there's an explicit `return` statement, not implicit returns. I think that's a logical and defensible distinction, not too specific an exception to carve out.

**Comment by mikecousins** [comment]
This conflicts pretty badly with this: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-string-refs.md

**Comment by gaearon** [comment]
To be clear, you can use callback refs just fine if you add `{` and `}`:

```js
<div ref={c => { this.hello = c }}>Hello, world.</div>
```

It’s kinda verbose so I agree it would be great if this option existed.

**Comment by gaearon** [comment]
Also, FYI, we removed this rule in Create React App because it is too inconvenient/surprising to React users who want to use callback refs (which by themselves are quite verbose so this makes it worse). Relevant discussion: https://github.com/facebookincubator/create-react-app/issues/528. Would be happy to provide more context on our decision if you’d like to chat!

**Comment by Kerumen** [comment]
Another way to do it, maybe a bit less verbose, is with parenthesis:

```javascript
<div ref={c => (this.hello = c)}>Hello, world.</div>
```

**Comment by vitorbal** [comment]
It looks like there's been a lot of support from the community lately for an option to ignore concise arrow-functions.
Since we originally closed this citing lack of support, I am hoping we can reach a consensus for a solution here, so I'm reopening this. If we can make it happen I'm willing to work on a PR for it.

**Comment by mysticatea** [comment]
I'm :-1:.
This request sounds the opposite of this rule's purpose.
Such change/option will make a hole to allow the `list.filter(x => x.code = 1)`-like mistakes.

Use of braces/parentheses makes clear the code.

**Comment by scottohara** [comment]
@mysticatea Please note my earlier comment that the use of braces/parens conflicts with other rules: https://github.com/eslint/eslint/issues/5150#issuecomment-191515093

**Comment by platinumazure** [comment]
@scottohara Have you looked at the options for no-extra-parens? I believe it is possible to suppress a warning of that rule for assignments in particular using an option. In that regard, both no-return-assign and no-extra-parens will happily allow use of parens to denote an intentional assignment here. (Compare with no-cond-assign.)

If it turns out that the current behavior does not match no-cond-assign, then perhaps we need to make a change. However, I'm otherwise :-1: with @mysticatea: the rule is doing exactly what it's supposed to here.

**Comment by mysticatea** [comment]
@scottohara `no-extra-parens` has the option to allow parentheses around such assignments.

```js
/*eslint no-return-assign: 2,
         no-extra-parens: [2, all, {returnAssign: false}],
         brace-style: 2,
         semi: 2
*/

somePromise.then(foo => (foo.bar = 1));
```

**Comment by scottohara** [comment]
OK, thanks @platinumazure / @mysticatea ; it looks like that is a new option that has been added since my original comment.

**Comment by btmills** [comment]
While I agree with the rest of the team here that the rule is behaving correctly and that disabling it defeats the purpose of the rule, given the choice between users disabling the rule entirely or just disabling the check for concise arrow functions, I'd rather they use part of the rule than none of it. Since so many people want it, I would support adding an option that would disable this check only arrow functions with expression bodies. I'd still like to see a proposal for what that option would look like, given the two current options "always" and "except-parens".

**Comment by kaicataldo** [comment]
Has this been resolved?

**Comment by nzakas** [comment]
With two :-1:s, and @btmills last comment, I think this can be closed.

**Comment by roderickObrist** [comment]
Hi All,

I may be a little late on this thread, but I find most of my angst with this rule can be determined based on the way the array is being used. For example:

``
// Good
someArray.forEach(val => str += val);
someObject.on('event', val => str += val);
someObject.then(val => str += val);

// Bad
someArray.map(val => str += val);
someObject.then(val => str += val)
  .then(hello => world);
``

To be specific, it would be a best of both worlds fix if:
1. There was a list of all Array methods that did not expect a return value and made an exception
2. If we can detect Promise chaining (.then().then().then()) made exception to the last arrow in the chain
3. In my experience, the return type does not matter in event handlers (.addEventListener, .on)

**Comment by remotezygote** [comment]
Looks like this got broken again. It no longer allows you to wrap these in parens. It yells as if the parens are not there.

**Comment by not-an-aardvark** [comment]
@remotezygote If something is broken, could you please file a separate issue and fill out the [bug report template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md)?

**Comment by remotezygote** [comment]
I currently can't spend the time on that at the immediate moment, but will when I can.

**Comment by zixuemeng** [comment]
write like this:
(buffer) => { data += buffer.toString(); }

**Comment by Kerumen** [comment]
@remotezygote see my comment here: https://github.com/eslint/eslint/issues/9359#issuecomment-332565278
Maybe, like me, you updated the Airbnb eslint config and thought eslint got broken.

**Comment by adrianmcli** [comment]
@remotezygote looks like it isn't only just you. Since this is the same problem as before, could we no just reopen this issue?

**Comment by not-an-aardvark** [comment]
This issue was a feature request, but it looks like you're reporting a bug, and it's not clear what behavior you're referring to or whether that behavior is intended. So I would recommend opening a new issue and filling out the [bug report template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md) to make it clearer what's going on.

**Comment by zenefits-brody** [comment]
Looking at the [doc for no-return-assign](https://eslint.org/docs/rules/no-return-assign), there is an default `except-parens`, which seems in favor of the workaround of wrapping the assignment expression in parens.
